woocommerce/woocommerce-ios
# [Mobile Payments] Fix IPP feedback survey banner display logic

Closes: #8992


## Description


This PR fixes various issues in the logic that determines if, and which, In Person Payments feedback survey banner to show the user. Full context: pdfdoF-2pO-p2

To fix these issues and show the correct banners, we've decided to patch up the iOS implementation to make it work as intended by:

1. Remove COD requirement for surveys 2 and 3.
2. Use Orders filtered to IPP-only for surveys 2 and 3 (instead of all WCPay Orders.)
3. Replace the filter which uses a server-localized string, `paymentMethodTitle` to use `paymentMethodID` instead, to avoid potential clashes with other plugins. (N.B. `paymentMethodID` is the same for IPP and web, so this isn’t sufficient to identify IPP transactions alone, but we use it in combination with the presence of receiptUrl, which is only set for IPP transactions within WCPay)

This PR is a step-by-step refactor, starting by fixing the tests and then progressively addressing the issues. A commit-by-commit review shouldn't be required, but there's some extra context in the commit messages if you choose to view it that way.

## Known issues
1. Locally stored orders are used, which means that getting the correct response requires the relevant orders to be loaded. In practice, this will often mean that the relevant orders must be in the first page (25) of orders.
2. Not all data is fetched from the same place, or at the same time. We constantly try to display the correct banner, so that results, in some cases, in one banner briefly being shown before the correct banner is shown. In practice, this seems very rare and I don't have any repro steps.

## Testing instructions


Testing fully requires several stores, but the steps are simple:

1. Open the Order list
2. Observe the correct banner is shown

### Scenarios worth testing:

**Using a US or CA based store**
No WCPay IPP orders (web orders are fine)
With COD enabled
Shows banner with text "Share your own experience or how you collect in-person payments."

Same store with COD turned off (you can toggle it in the Payments menu of the app)
Doesn't show a banner

Using a US or CA based store with 1-9 WCPay IPP payments in the last 30 days
With COD enabled or disabled
Shows banner with text "Rate your first in-person payment experience."

Using a US or CA based store with 10 or more WCPay IPP payments
With COD enabled or disabled
Shows banner with text "Tell us all about your experience with in-person payments."

Using a US or CA based store with 1-9 WCPay IPP payments, all more than 30 days old
With COD enabled or disabled
Doesn't show a banner

**Using another country based store**
Using one of the above banner-displaying stores, temporarily switched to a non-US in wc-admin
Doesn't show a banner

---
